egs2/TEMPLATE/enh_asr1/scripts/utils/eval_perm_free_error.py

```diff
@@ -0,0 +1,232 @@
+#!/usr/bin/env python3
+# encoding: utf-8
+
+# Copyright 2020 Johns Hopkins University (Xuankai Chang)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+import argparse
+import codecs
+import re
+import sys
+from collections import OrderedDict
+from pathlib import Path
+from typing import List
+
+import numpy as np
+import six
+from scipy.optimize import linear_sum_assignment
+
+sys.stdout = codecs.getwriter("utf-8")(sys.stdout.buffer)
+
+
+def get_parser():
+    parser = argparse.ArgumentParser(description="evaluate permutation-free error")
+    parser.add_argument(
+        "--num-spkrs", type=int, default=2, help="number of mixed speakers."
+    )
+    parser.add_argument(
+        "--results",
+        type=str,
+        nargs="+",
+        help="the scores between references and hypotheses, "
+        "in ascending order of references (1st) and hypotheses (2nd), "
+        "e.g. [r1h1, r1h2, r2h1, r2h2] in 2-speaker-mix case.",
+    )
+    parser.add_argument(
+        "--results-dir",
+        type=str,
+        help="the score dir. ",
+    )
+    return parser
+
+
+def convert_score(dic, num_spkrs=2) -> List[List[int]]:
+    ret = []
+    pat = re.compile(r"\d+")
+
+    for r_idx in range(num_spkrs):
+        ret.append([])
+        for h_idx in range(num_spkrs):
+            key = f"r{r_idx + 1}h{h_idx + 1}"
+
+            score = list(map(int, pat.findall(dic[key]["Scores"])))  # [c,s,d,i]
+            assert len(score) == 4  # [c,s,d,i]
+            ret[r_idx].append(score)
+
+    return ret
+
+
+def compute_permutation(old_dic, num_spkrs=2):
+    """Compute the permutation per utterance."""
+    all_scores, all_keys = [], list(old_dic.keys())
+    for scores in old_dic.values():  # compute error rate for each utt_id
+        all_scores.append(convert_score(scores, num_spkrs))
+    all_scores = np.array(all_scores)  # (B, n_ref, n_hyp, 4)
+
+    all_error_rates = np.sum(
+        all_scores[:, :, :, 1:4], axis=-1, dtype=np.float
+    ) / np.sum(
+        all_scores[:, :, :, 0:3], axis=-1, dtype=np.float
+    )  # (s+d+i) / (c+s+d), (B, n_ref, n_hyp)
+
+    min_scores, hyp_perms = [], []
+    for idx, error_rate in enumerate(all_error_rates):
+        row_idx, col_idx = linear_sum_assignment(error_rate)
+
+        hyp_perms.append(col_idx)
+        min_scores.append(np.sum(all_scores[idx, row_idx, col_idx], axis=0))
+
+    min_scores = np.stack(min_scores)
+
+    # Print results
```

Comment:
Is it better to leave the following comment-outed lines?

Reply:
I think so.
This script was written by @simpleoier. Do you think we can remove the commented part?

Reply:
We can just remove them.

Reply:
Yes, removing them is fine.

Reply:
OK, I will do that.